### PR TITLE
Nodelay

### PR DESCRIPTION
**Pull request by asvetlov**
Second approach to solve nodelay issue

### CODE REVIEW

File: aiohttp/parsers.py [hunk 1]
```diff
@@ -224,6 +225,22 @@ def __init__(self, transport, protocol, reader, loop):
...
         self._reader = reader
         self._loop = loop
+        self._tcp_nodelay = False
+        self._socket = transport.get_extra_info('socket')
+
+    @property
+    def tcp_nodelay(self):
+        return self._tcp_nodelay
+
+    def set_tcp_nodelay(self, value):
```
Comment (kxepal) [inline aiohttp/parsers.py:243]: Just curious, why not setter for`tcp_nodelay` property?

Comment (asvetlov) [inline aiohttp/parsers.py:243]: I thought about property's setter but found that syscall worth explicit function call.

Comment (kxepal) [inline aiohttp/parsers.py:243]: Ok, makes sense.

Comment (socketpair) [inline aiohttp/parsers.py:243]:
```
if value not in (1,0):
    raise ValueError('.....')
```

File: tests/test_web_response.py [hunk 2]
```diff
@@ -591,6 +591,48 @@ def test_prepare_calls_signal():
...
+
+    with mock.patch('aiohttp.web_reqrep.ResponseImpl'):
+        resp_impl = yield from resp.prepare(req)
+    resp_impl.transport.set_tcp_nodelay.assert_called_with(True)
+    resp.set_tcp_nodelay(False)
+    assert not resp.tcp_nodelay
+    resp_impl.transport.set_tcp_nodelay.assert_called_with(False)
+    resp.set_tcp_nodelay(True)
+    assert resp.tcp_nodelay
+    resp_impl.transport.set_tcp_nodelay.assert_called_with(True)
```
Comment (socketpair) [inline tests/test_web_response.py:632]: `True` ? maybe 1 or 0 ?

Comment (asvetlov) [inline tests/test_web_response.py:632]: `bool` is subclass of `int`

**Comment by asvetlov** [comment]
I'm finished.
Regular response and file uploading uses TCP_CORK to push all data in a single packet.
Streaming response and websocket use TCP_NODELAY to minimize latency.

Flow can be controlled by `set_tcp_nodelay()` and `set_tcp_cork()` calls. `tcp_nodelay` and `tcp_cork` states are mutually exclusive, case with neither nodelay nor cork is possible also (current aiohttp works exactly under this case).

I'll merge PR tomorrow. The only thing left is documentation.

**Comment by socketpair** [comment]
Is stream uncorked after writing last byte of response while in keep-alive mode ?

**Comment by socketpair** [comment]
Also, same technique should be applied to http client when it writes http request.

**Comment by socketpair** [comment]
I remember interesting technique in streaming over http and using chunked data. This cover not only EvenrSource http://www.w3schools.com/html/html5_serversentevents.asp but also videostreaming.

I mean that in these cases each chunk of chunked response contains complete data that should be delivered to browserwith minimal delay. Even when delay between chunks are long. So, in current implemntation, user should uncork manually after sending chunk and cork again before sending chunk. Does not it ? If yes, please document that example.

Also for websockets. Library should cork before sending frame and uncork afer sending. Is it so ?

Sorry, but code slightly complicated for me.

**Pull request merged**